agent/dns/router.go

```diff
@@ -0,0 +1,148 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package dns
+
+import (
+	"fmt"
+	"net"
+	"sync/atomic"
+	"time"
+
+	"github.com/armon/go-radix"
+	"github.com/hashicorp/go-hclog"
+	"github.com/miekg/dns"
+
+	"github.com/hashicorp/consul/agent/config"
+	"github.com/hashicorp/consul/agent/discovery"
+	"github.com/hashicorp/consul/agent/structs"
+)
+
+// TODO (v2-dns): metrics
+
+// RouterDynamicConfig is the dynamic configuration that can be hot-reloaded
+type RouterDynamicConfig struct {
+	ARecordLimit          int
+	DisableCompression    bool
+	EnableDefaultFailover bool // TODO (v2-dns): plumbing required for this new V2 setting. This is the agent configured default
+	EnableTruncate        bool
+	NodeMetaTXT           bool
+	NodeTTL               time.Duration
+	Recursors             []string
+	RecursorTimeout       time.Duration
+	RecursorStrategy      structs.RecursorStrategy
+	SOAConfig             SOAConfig
+	// TTLRadix sets service TTLs by prefix, eg: "database-*"
+	TTLRadix *radix.Tree
+	// TTLStrict sets TTLs to service by full name match. It Has higher priority than TTLRadix
+	TTLStrict      map[string]time.Duration
+	UDPAnswerLimit int
+}
+
+type SOAConfig struct {
+	Refresh uint32 // 3600 by default
+	Retry   uint32 // 600
+	Expire  uint32 // 86400
+	Minttl  uint32 // 0
+}
+
+// DiscoveryQueryProcessor is an interface that can be used by any consumer requesting Service Discovery results.
+// This could be attached to a gRPC endpoint in the future in addition to DNS.
+// Making this an interface means testing the router with a mock is trivial.
+type DiscoveryQueryProcessor interface {
+	QueryByName(*discovery.Query, discovery.Context) ([]*discovery.Result, error)
+	QueryByIP(net.IP, discovery.Context) ([]*discovery.Result, error)
+}
```

Comment:
Note: I modified the interface in the past day to distinguish between queries by name (normal DNS queries) and reverse queries (PTR queries). This was a suggestion from RFC comments and make more sense as I worked on the router.